### PR TITLE
Add context support (domain separation for signatures)

### PR DESCRIPTION
**Pull request by larabr**
The context is an abstraction we use to add domain separation to OpenPGP signatures.
 It's implemented by adding notation data to signatures, marked as critical, so that the resulting signature can only be verified by a verifier who expects the context to be present.
 
TODO:

- [x]  point to released openpgpjs

### CODE REVIEW

File: lib/message/context.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import type { SignaturePacket, RawNotation, PartialConfig } from '../openpgp';
+
+// Different contexts will affect the notation's value, not the name
+const CONTEXT_NOTATION_NAME = 'context@proton.ch';
+
+/**
+ * Translate a `context` string into an OpenPGP notation object, which can be signed as part of the message
+ * @param context - context identifier
+ */
+export const getNotationFromContext = (context: string): RawNotation => ({
```
Comment (twiss) [inline lib/message/context.ts:15]: Super nitpicky but this should be called `getNotationForContext` or `createNotationForContext`; `getNotationFromContext` would imply that you're taking the notation out of the context parameter somehow :)

File: lib/message/context.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        name === CONTEXT_NOTATION_NAME && new TextDecoder().decode(value) === context
+    ))
+);
+
+/**
+ * Signatures with context data can only be verified with a config that declares the context notation data as known
+ */
+export const getConfigForContextVerification = (config: PartialConfig) => ({
+    ...config,
+    knownNotations: [CONTEXT_NOTATION_NAME] // we can overwrite the field as we currently don't use any other known notations
```
Comment (twiss) [inline lib/message/context.ts:40]: We should just remove the default values, and then this comment can be removed

File: test/message/context.spec.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
+import { expect } from 'chai';
+import { verifyMessage, signMessage, generateKey, readSignature, readMessage, decryptMessage, encryptMessage, readKey } from '../../lib';
+import { VERIFICATION_STATUS } from '../../lib/constants';
+
+// verification without passing context should fail
+// verification passing wrong context should fail
+// verification of unsigned notation data should fail
```
Comment (twiss) [inline test/message/context.spec.ts:7]: Could you also add a test case of verifying a signature that doesn't have any context notation data, to check that it fails to verify (when a context is passed to `verifyMessage`)? I know it's almost the same as the unsigned one, but since it's the main case that we care about, I think it would be nice to have an explicit test for it (and not just one "hidden" as an edge case)

**Review by twiss — APPROVED** [review]
Seems good to me (once indeed we use the released OpenPGP.js)

File: lib/message/verify.js [hunk 4]
```diff
@@ -16,16 +17,17 @@ const { NOT_SIGNED, SIGNED_AND_VALID, SIGNED_AND_INVALID } = VERIFICATION_STATUS
  *           verified: Promise<Boolean>,
  *           signature: Promise<openpgp.signature.Signature>
  *         }
+ * @param {String} context - expected signature context
```
Comment (larabr) [inline lib/message/verify.js:20]:
```suggestion
 * @param {Object} context - context verification options
```

File: lib/message/context.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+     * Note: if the context is not required, but a different context is found in the signature, verification will always fail.
+     */
+    required: boolean,
+    requiredFrom?: undefined
+} | {
+    id: string,
+    required?: undefined
+    /**
+     * Signatures created after this date must include context info for verification to succeed (same as `expect: true`).
+     * For signatures created before the given date, the verification behavior is equivalent to `expect: false`
```
Comment (twiss) [inline lib/message/context.ts:38]:
```suggestion
     * Signatures created after this date must include context info for verification to succeed (same as `required: true`).
     * For signatures created before the given date, the verification behavior is equivalent to `required: false`
```

File: lib/message/context.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+     * */
+    requiredFrom: Date,
```
Comment (twiss) [inline lib/message/context.ts:40]: I would call this
```suggestion
    requiredAfter: Date,
```

File: lib/message/context.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    critical
+});
+
+/**
+ * Confirm context validity by finding the corresponding signature notation.
+ * NB: signature validity is not checked.
+ */
+export const isValidSignatureContext = (contextOptions: ContextVerificationOptions, signature: SignaturePacket) => {
+    const { id: contextID, required, requiredFrom } = contextOptions;
+    const isContextRequired = requiredFrom ? normalizeDate(requiredFrom) <= signature.created! : !!required;
```
Comment (twiss) [inline lib/message/context.ts:62]: And then this reads slightly more naturally as
```suggestion
    const { id: contextID, required, requiredAfter } = contextOptions;
    const isContextRequired = requiredAfter ? signature.created! > normalizeDate(requiredAfter) : !!required;
```

File: lib/message/context.ts [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    // `rawNotations` are always hashed (i.e. signed), otherwise OpenPGP's ignores them on parsing
+    const contextNotations = signature.rawNotations.filter(({ name }) => (name === CONTEXT_NOTATION_NAME));
+    const matchingContext = contextNotations.find(({ value }) => (new TextDecoder().decode(value) === contextID));
+
+    if (matchingContext) {
+        return true;
+    } else if (contextNotations.length > 0) {
+        // mismatching context are present
```
Comment (twiss) [inline lib/message/context.ts:71]:
```suggestion
        // mismatching context is present
```

File: lib/message/context.ts [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+ */
+import { SignaturePacket, RawNotation, PartialConfig } from '../openpgp';
+import { normalizeDate } from './utils';
+
+// Different contexts will affect the notation's value, not the name
+const CONTEXT_NOTATION_NAME = 'context@proton.ch';
+
+export interface ContextSigningOptions {
+    /** context identifier */
+    id: string,
```
Comment (twiss) [inline lib/message/context.ts:14]: I would call this `value` rather than `id`, even if it's a bit generic, it really could be anything in principle, not just an ID; e.g. in the case of Mail we could put some headers there, etc.

**Review by twiss — APPROVED** [review]
